File: apps/android/app/src/main/java/dev/crossbridge/android/network/RelayClient.kt
package dev.crossbridge.android.network

import io.ktor.client.HttpClient
import io.ktor.client.engine.okhttp.OkHttp
import io.ktor.client.plugins.websocket.WebSockets
import io.ktor.client.plugins.websocket.webSocketSession
import io.ktor.websocket.CloseReason
import io.ktor.websocket.Frame
import io.ktor.websocket.WebSocketSession
import io.ktor.websocket.close
import io.ktor.websocket.readText
import java.util.concurrent.CopyOnWriteArraySet
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONException
import org.json.JSONObject

enum class RelayConnectionState {
    DISCONNECTED,
    CONNECTING,
    CONNECTED,
    RECONNECTING,
    ERROR
}

class RelayClient(
    private val httpClient: HttpClient = HttpClient(OkHttp) {
        install(WebSockets)
    }
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val messageHandlers = CopyOnWriteArraySet<(JSONObject) -> Unit>()
    private val stateHandlers = CopyOnWriteArraySet<(RelayConnectionState) -> Unit>()

    @Volatile
    private var state: RelayConnectionState = RelayConnectionState.DISCONNECTED
    private var activeSession: WebSocketSession? = null
    private var currentUrl: String? = null
    private var receiveJob: Job? = null
    private var reconnectJob: Job? = null
    private var reconnectAttempts = 0
    private var manuallyDisconnected = false

    suspend fun connect(relayUrl: String) {
        withContext(Dispatchers.IO) {
            clearReconnect()
            closeActiveSession()
            currentUrl = relayUrl
            manuallyDisconnected = false
            reconnectAttempts = 0
            open(relayUrl, RelayConnectionState.CONNECTING)
        }
    }

    suspend fun send(message: JSONObject) {
        val session = activeSession ?: throw IllegalStateException("Relay socket is not connected.")
        session.outgoing.send(Frame.Text(message.toString()))
    }

    fun disconnect() {
        manuallyDisconnected = true
        clearReconnect()
        receiveJob?.cancel()
        val session = activeSession
        activeSession = null
        scope.launch {
            session?.close(CloseReason(CloseReason.Codes.NORMAL, "Disconnected"))
        }
        setState(RelayConnectionState.DISCONNECTED)
    }

    fun close() {
        disconnect()
        httpClient.close()
        scope.cancel()
    }

    fun onMessage(handler: (JSONObject) -> Unit): () -> Unit {
        messageHandlers.add(handler)
        return { messageHandlers.remove(handler) }
    }

    fun onStateChange(handler: (RelayConnectionState) -> Unit): () -> Unit {
        stateHandlers.add(handler)
        handler(state)
        return { stateHandlers.remove(handler) }
    }

    fun getState(): RelayConnectionState = state

    private suspend fun open(relayUrl: String, openingState: RelayConnectionState) {
        setState(openingState)
        try {
            val session = httpClient.webSocketSession(urlString = relayUrl)
            activeSession = session
            reconnectAttempts = 0
            setState(RelayConnectionState.CONNECTED)
            receiveJob = scope.launch {
                receiveLoop(session)
            }
        } catch (error: Throwable) {
            if (error is CancellationException) throw error
            setState(RelayConnectionState.ERROR)
            throw error
        }
    }

    private suspend fun receiveLoop(session: WebSocketSession) {
        try {
            for (frame in session.incoming) {
                if (frame is Frame.Text) {
                    dispatchMessage(frame.readText())
                }
            }
        } catch (error: Throwable) {
            if (error is CancellationException) throw error
            setState(RelayConnectionState.ERROR)
        } finally {
            val wasActiveSession = activeSession == session
            if (wasActiveSession) {
                activeSession = null
            }
            if (wasActiveSession && !manuallyDisconnected) {
                scheduleReconnect()
            }
        }
    }

    private fun dispatchMessage(raw: String) {
        val json = try {
            JSONObject(raw)
        } catch (_: JSONException) {
            setState(RelayConnectionState.ERROR)
            return
        }

        for (handler in messageHandlers) {
            handler(json)
        }
    }

    private fun scheduleReconnect() {
        val relayUrl = currentUrl
        if (relayUrl == null || reconnectAttempts >= RECONNECT_MAX_ATTEMPTS) {
            setState(RelayConnectionState.ERROR)
            return
        }

        reconnectAttempts += 1
        setState(RelayConnectionState.RECONNECTING)
        val delayMs = reconnectDelayMs(reconnectAttempts)

        reconnectJob = scope.launch {
            delay(delayMs)
            if (manuallyDisconnected) return@launch
            try {
                open(relayUrl, RelayConnectionState.RECONNECTING)
            } catch (_: Throwable) {
                scheduleReconnect()
            }
        }
    }

    private fun closeActiveSession() {
        receiveJob?.cancel()
        val session = activeSession
        activeSession = null
        if (session != null) {
            scope.launch {
                session.close(CloseReason(CloseReason.Codes.NORMAL, "Reconnecting"))
            }
        }
    }

    private fun clearReconnect() {
        reconnectJob?.cancel()
        reconnectJob = null
    }

    private fun setState(nextState: RelayConnectionState) {
        if (state == nextState) return
        state = nextState
        for (handler in stateHandlers) {
            handler(nextState)
        }
    }

    private companion object {
        const val RECONNECT_MAX_ATTEMPTS = Int.MAX_VALUE
    }
}

internal fun reconnectDelayMs(attempt: Int): Long {
    return when {
        attempt <= 1 -> 1_000L
        attempt == 2 -> 2_000L
        attempt == 3 -> 5_000L
        attempt == 4 -> 10_000L
        else -> 30_000L
    }
}

File: apps/windows/src/services/relayClient.ts
import type { RelayConnectionState } from "../types/pairing.js";

export type { RelayConnectionState };

type MessageHandler = (message: unknown) => void;
type StateHandler = (state: RelayConnectionState) => void;

const RECONNECT_DELAYS_MS = [1_000, 2_000, 5_000, 10_000];
const RECONNECT_MAX_DELAY_MS = 30_000;

export function reconnectDelayMs(attempt: number): number {
  if (attempt <= 0) return RECONNECT_DELAYS_MS[0];
  return RECONNECT_DELAYS_MS[attempt - 1] ?? RECONNECT_MAX_DELAY_MS;
}

export class RelayClient {
  private socket?: WebSocket;
  private url?: string;
  private state: RelayConnectionState = "disconnected";
  private manuallyDisconnected = false;
  private reconnectAttempts = 0;
  private reconnectTimer?: ReturnType<typeof setTimeout>;
  private readonly messageHandlers = new Set<MessageHandler>();
  private readonly stateHandlers = new Set<StateHandler>();

  connect(url: string): Promise<void> {
    this.clearReconnectTimer();
    this.closeSocket();
    this.url = url;
    this.manuallyDisconnected = false;
    this.reconnectAttempts = 0;
    return this.open(url, "connecting");
  }

  disconnect(): void {
    this.manuallyDisconnected = true;
    this.clearReconnectTimer();
    this.closeSocket();
    this.setState("disconnected");
  }

  send(message: unknown): void {
    if (!this.socket || this.socket.readyState !== WebSocket.OPEN) {
      throw new Error("Relay socket is not connected.");
    }
    this.socket.send(JSON.stringify(message));
  }

  onMessage(handler: MessageHandler): () => void {
    this.messageHandlers.add(handler);
    return () => {
      this.messageHandlers.delete(handler);
    };
  }

  onStateChange(handler: StateHandler): () => void {
    this.stateHandlers.add(handler);
    handler(this.state);
    return () => {
      this.stateHandlers.delete(handler);
    };
  }

  getState(): RelayConnectionState {
    return this.state;
  }

  private open(url: string, state: RelayConnectionState): Promise<void> {
    this.setState(state);

    return new Promise((resolve, reject) => {
      const socket = new WebSocket(url);
      this.socket = socket;
      let settled = false;

      const resolveOnce = (): void => {
        if (settled) return;
        settled = true;
        resolve();
      };

      const rejectOnce = (error: Error): void => {
        if (settled) return;
        settled = true;
        reject(error);
      };

      socket.addEventListener("open", () => {
        if (this.socket !== socket) return;
        this.reconnectAttempts = 0;
        this.setState("connected");
        resolveOnce();
      }, { once: true });

      socket.addEventListener("message", (event) => {
        if (this.socket !== socket) return;
        try {
          this.emitMessage(JSON.parse(String(event.data)));
        } catch {
          this.setState("error");
        }
      });

      socket.addEventListener("error", () => {
        if (this.socket !== socket) return;
        if (this.manuallyDisconnected || !this.url) {
          this.setState("error");
        }
        rejectOnce(new Error("Relay connection failed."));
      }, { once: true });

      socket.addEventListener("close", () => {
        const wasActiveSocket = this.socket === socket;
        if (wasActiveSocket) {
          this.socket = undefined;
        }

        if (!wasActiveSocket) {
          return;
        }

        if (this.manuallyDisconnected) {
          this.setState("disconnected");
          return;
        }

        this.scheduleReconnect();
        rejectOnce(new Error("Relay connection closed."));
      });
    });
  }

  private scheduleReconnect(): void {
    if (!this.url) {
      this.setState("error");
      return;
    }

    this.reconnectAttempts += 1;
    this.setState("reconnecting");
    const delay = reconnectDelayMs(this.reconnectAttempts);
    this.clearReconnectTimer();

    this.reconnectTimer = setTimeout(() => {
      if (!this.url || this.manuallyDisconnected) return;
      void this.open(this.url, "reconnecting").catch(() => {
        if (!this.socket && !this.manuallyDisconnected && this.url) {
          this.scheduleReconnect();
        }
      });
    }, delay);
  }

  private closeSocket(): void {
    const socket = this.socket;
    this.socket = undefined;
    if (socket && socket.readyState !== WebSocket.CLOSED) {
      socket.close();
    }
  }

  private clearReconnectTimer(): void {
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = undefined;
    }
  }

  private setState(nextState: RelayConnectionState): void {
    if (this.state === nextState) return;
    this.state = nextState;
    for (const handler of this.stateHandlers) {
      handler(nextState);
    }
  }

  private emitMessage(message: unknown): void {
    for (const handler of this.messageHandlers) {
      handler(message);
    }
  }
}
